### ISSUE TITLE
Release 4.5.2 to address CVE-2021-44228

### ISSUE THREAD

**Issue by iloveeclipse** [spotbugs/spotbugs#1868]
We should release 4.5.2 version to quickly address CVE-2021-44228.
The PR #1865 is prerequisite.

Note: there is no known exploit using spotbugs possible yet, but since we ship and use log4j in affected version, and there are web services that run spotbugs as a part of the build pipeline, an attacker could manage to craft a pull request that contains the Java code that would somehow manage the bad pattern be part of the message logged by spotbugs.

**Comment by iloveeclipse** [comment]
@KengoTODA : the build fails with this error below: any idea?
```
> Task :closeSonatypeStagingRepository FAILED
FAILURE: Build failed with an exception.


Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0.
* What went wrong:

Execution failed for task ':closeSonatypeStagingRepository'.
You can use '--warning-mode all' to show the individual deprecation warnings and determine if they come from your own scripts or plugins.
> Staging repository is still transitioning after defined time. Consider its increment. StagingRepository(id=comgithubspotbugs-1253, state=open, transitioning=true)
```

**Comment by KengoTODA** [comment]
> Staging repository is still transitioning after defined time. Consider its increment. StagingRepository(id=comgithubspotbugs-1253, state=open, transitioning=true)

It seems that the Sonatype Nexus (oss.sonatype.org) took a long time to handle our request. The deployment itself has been succeeded.
I will check the diff, and release pushed artifact if found no problem.

**Comment by iloveeclipse** [comment]
Please approve and merge #1871 first

**Comment by KengoTODA** [comment]
It seems that Apache Foundation released 2.16.0, I'll check its release note.

https://github.com/apache/logging-log4j2/releases/tag/rel%2F2.16.0

**Comment by iloveeclipse** [comment]
see https://lists.apache.org/thread/6g0ls1cc9htn52hr0kkbo7dds5ts3rb5

**Comment by iloveeclipse** [comment]
I assume we could pick this release too.

**Comment by KengoTODA** [comment]
checked, thanks. Is it better to release our 4.5.2 with 2.16.0? It is possible for now.

**Comment by iloveeclipse** [comment]
Please update.

**Comment by iloveeclipse** [comment]
@KengoTODA : I think we should release maven and gradle plugins ASAP too, once core is released. WDYT? Would you push that?

**Comment by KengoTODA** [comment]
log4j is packaged only into the zip/tarball distribution, so Maven/Gradle plugins are not affected by this vulnerability.

You can see that runtime dependency of 4.5.1 release does not have log4j2 at:
https://mvnrepository.com/artifact/com.github.spotbugs/spotbugs/4.5.1

**Comment by KengoTODA** [comment]
refer to https://github.com/spotbugs/spotbugs/pull/790

**Comment by iloveeclipse** [comment]
I'm not sure if the data on maven repository is correct.
SpotBugs uses slf4j, and that one requires some binding. Once the log4j binding is used, the runtime is affected.

**Comment by KengoTODA** [comment]
I mean, SpotBugs core depends on `slf4j-api` only. Who depends on SpotBugs core needs to prepare the binding by their own.
For instance, Maven plugin uses [slf4j-simple](https://github.com/spotbugs/spotbugs-maven-plugin/blob/6f9486f9991ad8c78f7afc17a08507ab52613f81/pom.xml#L209). Gradle plugin also uses [slf4j-simple](https://github.com/spotbugs/spotbugs-gradle-plugin/blob/66c5859c242ec2cba643992e8eaad3a81851d640/src/main/groovy/com/github/spotbugs/snom/SpotBugsBasePlugin.java#L102).

> There is no way to say slf4j to require some specific version of Log4j?

There is no way. `slf4j-api` has no dependency on log4j's API and bindings.

**Comment by clh15683** [comment]
The released version contains both 2.15.0 and 2.16.0 jars for log4j. By classpath order log4jcore-2.15 and log4j-api 2.16 are used. While this fixes the CVE, ist should be fixed to reduce the jar size and use consistent versions.

**Comment by iloveeclipse** [comment]
@clh15683 : good find!
In fact, https://github.com/spotbugs/spotbugs/releases/download/4.5.2/eclipsePlugin.zip contains both libraries.

@KengoTODA : this happens if one builds incrementally - gradle build seem not to properly cleanup previous data.  Not sure how you build the package, but ideally one should run `git clean -fdx` to make sure the working tree is clean.

**Comment by KengoTODA** [comment]
The release build does not use cache, so `./gradlew clean` is needless and the `build` directory is always empty at the beginning.

https://github.com/spotbugs/spotbugs/blob/4.5.2/.github/workflows/release.yml

We use dependency cache in `actions/setup-java` but [it caches dependencies only](https://github.com/actions/setup-java/blob/v2.4.0/src/cache.ts) so it won't be affected.

I'll check the build process itself. The eclipse plugin probably have two or more dependencies on log4j, then [dependency constraints](https://blog.gradle.org/log4j-vulnerability) could be help.

**Comment by KengoTODA** [comment]
It seems that the `:spotbugs:copyLibsForEclipse` task and `:eclipsePlugin:copyLibsForEclipse` task copy these libraries. You can confirm it by `git clean -fdx && ./gradlew copyLibsForEclipse`.

@iloveeclipse Do we need to include logging framework's API and bindings? More specifically, they are jar files as follows:

- log4j-api-2.14.0.jar
- log4j-api-2.14.1.jar
- log4j-api-2.15.0.jar
- log4j-api-2.16.0.jar
- log4j-core-2.14.0.jar
- log4j-core-2.14.1.jar
- log4j-core-2.15.0.jar
- log4j-core-2.16.0.jar
- log4j-slf4j18-impl-2.14.0.jar
- log4j-slf4j18-impl-2.14.1.jar
- log4j-slf4j18-impl-2.15.0.jar
- log4j-slf4j18-impl-2.16.0.jar
- slf4j-api-1.8.0-beta4.jar

I expect that Eclipse itself will provide API and binding in runtime. If so, we can simply exclude these libraries from the target of the `copyLibsForEclipse` task.

**Comment by KengoTODA** [comment]
BTW a permanent solution is removing this task, and it means that we need to resolve dependencies from Maven Central. It's technically possible but I have one blocker, so I [posted to Eclipse Community Forum](https://www.eclipse.org/forums/index.php/t/1109122/) to ask a question at this October.

If there is no answer in that forum, I will find another way to download and extract [org.eclipse.pde.build](https://mvnrepository.com/artifact/org.eclipse.pde/org.eclipse.pde.build/3.9.200).

**Comment by iloveeclipse** [comment]
Not sure why there are so many versions? The "copy" tasks only copy declared dependencies. 
I assume only one version for each library should be available at all?
Where are all the older versions coming from?

I think slf4j-api-1.8.0-beta4.jar is required, but not sure regarding log4j, I have to debug. In any case, everything that is not required to be in the plugin, must be declared as dependency in manifest.

**Comment by KengoTODA** [comment]
> Not sure why there are so many versions? The "copy" tasks only copy declared dependencies.
> I assume only one version for each library should be available at all?
> Where are all the older versions coming from?

I believe the direct reason is that this task need to copy dependencies in `testRuntimeClasspath`, to support Eclipse runs unit test. So it copies not only log API but also log binding defined at:

https://github.com/spotbugs/spotbugs/blob/49fc6f68d6a76dcf931fdbc93569e82c65a931c9/spotbugs/build.gradle#L99

I will update this line, and regenerate manifest files for the Eclipse plugin.

**Comment by cypher256** [comment]
Log4j 2.16 CVSS 7.5 : Fixed in Log4j 2.17
https://logging.apache.org/log4j/2.x/security.html#CVE-2021-45105

**Comment by iloveeclipse** [comment]
@cypher256 : thanks. I've created https://github.com/spotbugs/spotbugs/issues/1882

**Comment by KengoTODA** [comment]
[SpotBugs 4.5.3](https://github.com/spotbugs/spotbugs/releases/tag/4.5.3) has been released, which depends on log4j 2.17.1.
